File: Primes.java
package jon;

import java.util.Vector;
import java.util.Iterator;

public class Primes
{
    
    public static void main(String[] args) 
    {
        if (args.length < 2) {
            System.out.println("You need to give a limit and container type, yo!");
            System.exit(1);
        }
        
        long startTime = System.currentTimeMillis();

        boolean doPrint = args.length > 2 && args[2].equals("-p");
        int lim = Integer.parseInt(args[0]);
        
        if (args[1].equals("vector"))
            withVector(lim, doPrint);
        else
            withArray(lim, doPrint);
        
        System.out.printf("Time: %dms\n", System.currentTimeMillis() - startTime);
    }

    private static void withArray(int lim, boolean doPrint) {
        int[] primes = new int[lim/10];
        primes[0] = 2;
        int n = 3, endIdx = 1, p, pIdx = 0;

        while (n <= lim) {
            p = primes[pIdx];
            if (endIdx == primes.length) {
                int[] newArray = new int[primes.length * 2];
                System.arraycopy(primes, 0, newArray, 0, primes.length);
                primes = newArray;
            }
            
            if (p > (n / p)) {
                pIdx = 0;
                primes[endIdx++] = n;
                n++;
            } else if (n % p == 0) {
                pIdx = 0;
                n++;
            } else
                pIdx++;
        }

        if (doPrint) {
            System.out.print("[");
            if (endIdx > 20) {
                for (pIdx = 0; pIdx < 10; pIdx++)
                    System.out.printf("%d ", primes[pIdx]);
                System.out.print("... ");
                for (pIdx = endIdx - 10; pIdx < endIdx - 1; pIdx++)
                    System.out.printf("%d ", primes[pIdx]);
            } else {
                for (pIdx = 0; pIdx < primes.length - 1; pIdx++)
                    System.out.printf("%d ", primes[pIdx]);
            }
            System.out.printf("%d]\n", primes[pIdx]);
            System.out.printf("Total: %d\n", endIdx);
        }
    }

    private static void withVector(int lim, boolean doPrint) {
        Vector<Integer> primes = new Vector<Integer>();
        primes.add(2);
        int n = 3, pIdx = 0, p;
        
        while (n <= lim) {
            p = primes.get(pIdx);
            if (p > (n / p)) {
                pIdx = 0;
                primes.add(n);
                n++;
            } else if ((n % p) == 0) {
                pIdx = 0;
                n++;
            } else
                pIdx++;
        }

        if (doPrint) {
            System.out.print("[");
            if (primes.size() > 20) {
                for (pIdx = 0; pIdx < 10; pIdx++)
                    System.out.printf("%d ", primes.get(pIdx));
                System.out.print("... ");
                for (pIdx = primes.size() - 10; pIdx < primes.size() - 1; pIdx++)
                    System.out.printf("%d ", primes.get(pIdx));
            } else {
                for (pIdx = 0; pIdx < primes.size() - 1; pIdx++)
                    System.out.printf("%d ", primes.get(pIdx));
            }
            System.out.printf("%d]\n", primes.lastElement());
            System.out.printf("Total: %d\n", primes.size());
        }
    }
                                
}

File: primes.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/time.h>

int main(int argc, char *argv[]) {
  if (argc < 2) {
    fprintf(stderr, "You need to give a limit, yo!\n");
    return 1;
  }

  int do_print = (argc == 3 && !strcmp(argv[2], "-p"));
  
  struct timeval the_time;
  gettimeofday(&the_time, NULL);
  suseconds_t start_time = the_time.tv_usec;

  int lim = atoi(argv[1]);
  int alloc_bytes = lim/10 * sizeof(int), 
    alloc_ints = lim/10;
  
  int *primes = malloc(alloc_bytes);
  int n = 3;
  *primes = 2;
  int end_idx = 1;

  int p_idx = 0, p;
  while (n <= lim) {
    if (end_idx == alloc_ints) {
      alloc_bytes += alloc_bytes;
      alloc_ints += alloc_ints;
      
      primes = realloc(primes, alloc_bytes);
      if (primes == NULL) {
        fprintf(stderr, "ERROR: Reallocation failed!\n");
        return 1;
      }
    }

    p = primes[p_idx];
    if (p > (n / p)) { 
      p_idx = 0;
      primes[end_idx++] = n;
      n++;
    } else if (! (n % p)) { 
      p_idx = 0;
      n++;
    } else
      p_idx++;
  }

  if (do_print) {
    printf("[");
    if (end_idx > 20) {
      for (p_idx = 0; p_idx < 10; ++p_idx)
        printf("%i ", primes[p_idx]);
      printf("... ");
      for (p_idx = end_idx - 10; p_idx < end_idx - 1; p_idx++)
        printf("%i ", primes[p_idx]);
    } else {
      for (p_idx = 0; p_idx < end_idx - 1; p_idx++)
        printf("%i ", primes[p_idx]);
    }
    printf("%i]\n", primes[p_idx]);
    printf("Total: %i\n", end_idx);
  }

  gettimeofday(&the_time, NULL);
  suseconds_t end_time = the_time.tv_usec;
  printf("Time: %ims\n", (end_time - start_time) / 1000);

  return 0;
}
